### DISCUSSION TITLE
Delete statements take 1s per hit

### DISCUSSION THREAD

**Discussion by Bechma** [sqlalchemy/sqlalchemy#7076] [Usage Questions] [answered]
Hello, I was porting some synchronous code to asynchronous. Everything went fine except for one function.
It was taking too much time, so I decided to profile both synchronous and asynchronous code. The problem was in literally one statement:

```python
# child_tickets holds an array of declarative_base classes reference
for table in child_tickets:
    # The profiler shows 98% of the total execution time was expended here. 4837.6s and 4825 hits
    await session.execute(delete(table).where(table.id.in_(ticket.id for ticket in ticket_list)))
await session.execute(upsert_tickets, [tickets.upsert_ticket(i) for i in ticket_list])
session.add_all(obj)
await session.commit()
```

Due to this, the process went from 25m of execution to more than 2h.
Am I deleting the tables wrongly? Why python goes to 100% CPU usage there?

Thanks in advance

Database is Postgresql 12, Python 3.9 and SQLAlchemy 1.4.25

**Comment by zzzeek** [comment, marked as answer]
You would need to look at the profiler result and see if the time is spent either in Python computation or in database IO, as well as if some usage of the ORM is leading to lots of unwanted queries occurring, such as lazy loads for each record you place in the statement.

It would be best to work with synchronous code for profilng, if possible.  then , go through the steps at https://docs.sqlalchemy.org/en/14/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application first viewing if unwanted SQL statements are emitting, then looking to see how much time is spent waiting on database IO vs. how much is in python computation.   if this list of records is in the tens of thousands or millions, the result is not that surprising, ORM objects should not be used in these scales, I would use plain dictionaries / tuples instead with Core -only.

**Comment by CaselIT** [comment]
Another suggestion is to split the statement that takes the longest time so you can find what's actually taking time:
`await session.execute(delete(table).where(table.id.in_(ticket.id for ticket in ticket_list)))`
becomes
```py
items = [ticket.id for ticket in ticket_list]
in_ = table.id.in_(items)
stmt = delete(table).where(in_)
await session.execute(stmt)
```

While writing the above I noticed that you are passing a generator to the `in_()`: `table.id.in_(ticket.id for ticket in ticket_list)`
I don't think that's supported

**Comment by Bechma** [comment]
Thanks for the replies, after some tries, I managed to reduce a little bit the time taken by the delete statement adding an expunge_all after the commit. However, there is still a memory leak problem during the execution. Even though I am invalidating the objects in the session, when the process finish, I have around 3GB of memory that I have to manually call gc.collect() to free up 2.6GB... Do you have any idea why the garbage collector is not fired up when expunging the objects?

**Reply by CaselIT** [reply]
With the information you provided is not easy to guess what's happening.
Can you provide an extended version of the snippet, and if you can both in the sync ( that was working) and async forms?
Thanks

**Reply by Bechma** [reply]
Async version:

```python
# engine and session:
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession, async_scoped_session
from sqlalchemy.orm import sessionmaker
from asyncio import current_task

engine = create_async_engine(db_connstring, pool_pre_ping=True)
session = async_scoped_session(
    sessionmaker(bind=engine, expire_on_commit=False, class_=AsyncSession),
    scopefunc=current_task)
###
###
###
upsert_tickets = insert(orm.Tickets).on_conflict_do_update(
    constraint=orm.Tickets.__table__.primary_key,
    set_=orm.Tickets.__table__.c
)
for i in asyncio.as_completed([processing.parse_tickets(i) for i in total_tickets]):
    tickets, tickets_history = await i
    connection = await session.connection()
    await connection.execution_options(schema_translate_map={None: "schema1"})
    for table in child_tickets:
        await session.execute(delete(table).where(table.id.in_([ticket.id for ticket in tickets])))
    await session.execute(upsert_tickets, [processing.tickets.upsert_ticket(i) for i in tickets])
    session.add_all(tickets_history)
    await session.commit()
    session.expunge_all()
```

And sync version:
```python
# engine and session:
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker, scoped_session

engine = create_engine(db_connstring, executemany_values_page_size=10000, executemany_batch_page_size=500)
session = scoped_session(sessionmaker(bind=engine))
###
###
###
with multiprocessing.Pool() as pool:
    for tickets, obj in pool.imap_unordered(processing.parse_tickets, total_tickets):
        session.connection(execution_options={"schema_translate_map": {None: "schema1"}})
        for table in child_tickets:
            session.execute(delete(table).where(table.id.in_([ticket.id for ticket in tickets])))
        for ticket in tickets:
            ticket_upsert = upsert_ticket(ticket)
            session.execute(insert(orm.Tickets).values(**ticket_upsert).on_conflict_do_update(
                constraint=orm.Tickets.__table__.primary_key,
                set_=ticket_upsert
            ))
        session.bulk_save_objects(obj)
        session.commit()
```

**Reply by zzzeek** [reply]
"I have around 3GB of memory that I have to manually call gc.collect() to free up 2.6GB.."  that's not a memory leak if gc.collect() is able to clear it.   do you have automatic gc turned off ?

**Reply by Bechma** [reply]
You were right zzzeek, after much more profiling, I detect the problem were somewhere else, not in sqlalchemy.

Thanks for the help
